### ISSUE TITLE
ACR url should deploy with Registry source as Azure Container Registry

### ISSUE THREAD

**Issue by rawilder** [Azure/webapps-deploy#136]
When using the following action config, it will deploy to the webapp instance configured with the Registry source as "Private Registry" which makes the deployment fail requiring a username and password. With managed identity, it should be fine setting it to Azure Container Registry so the existing permissions allow a smooth experience.

Workaround: Change the Registry source to Azure Container Registry after the first deploy with the matching container, then future actions seem to apply that. 

```
      - name: Azure WebApp
        uses: Azure/webapps-deploy@v2
        with:
          # Name of the Azure Web App
          app-name: ${{ env.APP_NAME }}
          # Applies to Web App Containers only: Specify the fully qualified container image(s) name. For example, 'myregistry.azurecr.io/nginx:latest' or 'python:3.7.2-alpine/'. For multi-container scenario multiple container image names can be provided (multi-line separated)
          images: ${{ secrets.ACR_LOGIN_SERVER }}/myapp:${{ github.sha }}
          # Enter the start up command. For ex. dotnet run or dotnet run
          startup-command: /home/app/start.sh
```

**Comment by kf6kjg** [comment]
If I read the README correctly this behavior has been pawned off to the [`azure/appservice-settings` action](https://github.com/Azure/appservice-settings). It'd be nice to have this in a one-stop-shop though.

**Comment by kf6kjg** [comment]
After posting that yesterday I kept working on it. You are right.  However there's a gotcha that might be the reason this action doesn't have this feature.

I wound up replacing this action with roughly, though not exactly, equivalent Azure CLI commands:

1. `az webapp config container set` sets the `--docker-custom-image-name` and `--docker-registry-server-url` for the ACR.  But that needs to have the App Service, or slot if you are using one like I am, be granted access to the ACR. 
2. Granting access to the ACR can be done via the `az role assignment create` with the `AcrPull` role - which can be specified using the ID instead of the same to save on permissions requirements.  You need to feed this command with the principle ID of the service or slot along with the ID of the ACR. But here is where the gotcha lies.

The `az role assignment create` command requires the active user to have some pretty hefty permissions. I'm right now in the process of seeing if my bosses will allow the automated deployment user that is being used by by GitHub workflow to have those permissions. For reference the error returned was that the client "... does not have authorization to perform action 'Microsoft.Authorization/roleAssignments/write' over scope ... or the scope is invalid." I'm reasonably sure the scope is valid, having pulled the information directly from `az acr show`.  

This means that the Azure auth in use on a 3rd party service aka GHActions, accessible by other 3rd party arbitrary code aka all the actions in use by my script, would be able to do arbitrary role assignments. Not sure I can get that kind of security risk pushed through - and might explain why this Action doesn't attempt to accomplish this.

**Comment by rawilder** [comment]
This is very helpful! Thank you. I wonder if there'd be a way to work around if if the role just already exists and there's no need to grant access on the deployment. Like `if app has access already then just use acr type else use registry url, username, and password`.

**Comment by Whyeasy** [comment]
> 
> 
> This is very helpful! Thank you. I wonder if there'd be a way to work around if if the role just already exists and there's no need to grant access on the deployment. Like `if app has access already then just use acr type else use registry url, username, and password`.

This would be great to have. Currently I'm running in to an issue and this would solve it. I create my infrastructure via Terraform and have the managed identity set and assigned the `acrPull` role in the ACR resource. To make the app service authenticate with ACR and use the managed identity, you need to set an extra value. This value is overwritten with this action as it's part of the `site_config` in the Terraform resource. As a result it can't pull the image.

**Comment by sgollapudi77** [comment]
> If you are referring to this [action](https://github.com/Azure/webapps-deploy#configure-web-app-private-registry-credentials), then this is not the solution I am talking about.
> 
> Azure app service deployment configurations have a way to configure implicit auth with `Registry source: Azure Container Regsitry`. It does not require username/password.
> 
> So I suppose the best way to look at this is enhancement request since it can technically be done, it just requires extra steps that shouldn't strictly be necessary.

Hi @rawilder, we understand that there are manual steps needed to pull the image from registry, but these steps need to be done only once and the subsequent deployments will work fine.
However we'll try to look into this enhancement and will add it to our backlog.
